From 91af6e5cb7bff4034ef5d3eb5acd2a7eb2c4691e Mon Sep 17 00:00:00 2001
From: Barry Wardell <barry.wardell@gmail.com>
Date: Mon, 6 Feb 2012 16:52:49 +0000
Subject: [PATCH] CarpetLib: Improve prolongation with single precision.

* Allow larger errors
* Disable 11th order as single precision isn't good enough for it to be accurate.
---
 Carpet/CarpetLib/src/prolongate_3d_rf2.cc |   68 ++++++++++++++++++++++------
 1 files changed, 53 insertions(+), 15 deletions(-)

diff --git a/Carpet/CarpetLib/src/prolongate_3d_rf2.cc b/Carpet/CarpetLib/src/prolongate_3d_rf2.cc
index 246cbf8..9f76b43 100644
--- a/Carpet/CarpetLib/src/prolongate_3d_rf2.cc
+++ b/Carpet/CarpetLib/src/prolongate_3d_rf2.cc
@@ -76,13 +76,15 @@ namespace CarpetLib {
       for (int n=0; n<=ORDER; ++n) {
         RT res = RT(0.0);
         for (ptrdiff_t i=imin; i<imax; ++i) {
-          RT const x = RT(CCTK_REAL(i) - 0.5);
+          RT const x = RT(i) - RT(0.5);
           RT const y = ipow (x, n);
           res += get(i) * y;
         }
         RT const x0 = RT(0.0);
         RT const y0 = ipow (x0, n);
-        if (not (good::abs (res - y0) < 1.0e-12)) {
+        // Allow losing 3 digits:
+        CCTK_REAL const eps = RT(1.0e+3) * numeric_limits<RT>::epsilon();
+        if (not (good::abs (res - y0) < eps)) {
           RT rt;
           ostringstream buf;
           buf << "Error in prolongate_3d_rf2::coeffs_3d_rf2\n"
@@ -102,22 +104,23 @@ namespace CarpetLib {
   
   
   
-#define TYPECASE(N,RT)                          \
-                                                \
+#define COEFFS1D_1(RT)                          \
   template<>                                    \
   const RT coeffs1d<RT,1>::coeffs[] = {         \
     +1 / RT(2.0),                               \
     +1 / RT(2.0)                                \
-  };                                            \
-                                                \
+  };
+
+#define COEFFS1D_3(RT)                          \
   template<>                                    \
   const RT coeffs1d<RT,3>::coeffs[] = {         \
     -1 / RT(16.0),                              \
     +9 / RT(16.0),                              \
     +9 / RT(16.0),                              \
     -1 / RT(16.0)                               \
-  };                                            \
-                                                \
+  };
+
+#define COEFFS1D_5(RT)                          \
   template<>                                    \
   const RT coeffs1d<RT,5>::coeffs[] = {         \
     +  3 / RT(256.0),                           \
@@ -126,8 +129,9 @@ namespace CarpetLib {
     +150 / RT(256.0),                           \
     - 25 / RT(256.0),                           \
     +  3 / RT(256.0)                            \
-  };                                            \
-                                                \
+  };
+
+#define COEFFS1D_7(RT)                          \
   template<>                                    \
   const RT coeffs1d<RT,7>::coeffs[] = {         \
       -   5 / RT(2048.0),                       \
@@ -138,8 +142,9 @@ namespace CarpetLib {
       - 245 / RT(2048.0),                       \
       +  49 / RT(2048.0),                       \
       -   5 / RT(2048.0)                        \
-  };                                            \
-                                                \
+  };
+
+#define COEFFS1D_9(RT)                          \
   template<>                                    \
   const RT coeffs1d<RT,9>::coeffs[] = {         \
       +   35 / RT(65536.0),                     \
@@ -152,8 +157,9 @@ namespace CarpetLib {
       + 2268 / RT(65536.0),                     \
       -  405 / RT(65536.0),                     \
       +   35 / RT(65536.0)                      \
-  };                                            \
-                                                \
+  };
+
+#define COEFFS1D_11(RT)                         \
   template<>                                    \
   const RT coeffs1d<RT,11>::coeffs[] = {        \
       -    63 / RT(524288.0),                   \
@@ -170,9 +176,32 @@ namespace CarpetLib {
       -    63 / RT(524288.0)                    \
   };
   
+#ifdef CCTK_REAL_PRECISION_4
+#define TYPECASE(N,RT) \
+  COEFFS1D_1(RT)       \
+  COEFFS1D_3(RT)       \
+  COEFFS1D_5(RT)       \
+  COEFFS1D_7(RT)       \
+  COEFFS1D_9(RT)       \
+  COEFFS1D_11(RT)
+#else
+#define TYPECASE(N,RT) \
+  COEFFS1D_1(RT)       \
+  COEFFS1D_3(RT)       \
+  COEFFS1D_5(RT)       \
+  COEFFS1D_7(RT)       \
+  COEFFS1D_9(RT)
+#endif
+/* Only define up to 9th order prolongation for single precision */
 #define CARPET_NO_COMPLEX
 #include "typecase.hh"
 #undef TYPECASE
+#undef COEFFS1D_1
+#undef COEFFS1D_3
+#undef COEFFS1D_5
+#undef COEFFS1D_7
+#undef COEFFS1D_9
+#undef COEFFS1D_11
     
     
     
@@ -181,13 +210,22 @@ namespace CarpetLib {
     {
       DECLARE_CCTK_ARGUMENTS;
       
-#define TYPECASE(N,RT)                          \
+#ifdef CCTK_REAL_PRECISION_4
+#  define TYPECASE(N,RT)                        \
+      coeffs1d<RT,1>::test();                   \
+      coeffs1d<RT,3>::test();                   \
+      coeffs1d<RT,5>::test();                   \
+      coeffs1d<RT,7>::test();                   \
+      coeffs1d<RT,9>::test();
+#else
+#  define TYPECASE(N,RT)                        \
       coeffs1d<RT,1>::test();                   \
       coeffs1d<RT,3>::test();                   \
       coeffs1d<RT,5>::test();                   \
       coeffs1d<RT,7>::test();                   \
       coeffs1d<RT,9>::test();                   \
       coeffs1d<RT,11>::test();
+#endif
 #define CARPET_NO_COMPLEX
 #define CARPET_NO_INT
 #include "typecase.hh"
-- 
1.7.7.4

